off-chain workers htg - transactions (#684)
substrate-developer-hub/substrate-docs: v3/how-to-guides/04-ocw/a-transactions/index.mdx

---
title: Making Transactions
slug: /how-to-guides/v3/ocw/transactions
keywords:
  - off-chain worker
  - ocw
  - transactions
version: '3.0'
section: how to guides
category: ocw
---

<Objectives data={[
  {
    title: 'Goal',
    description: `Learn how to perform the following actions from off-chain workers:
- Signed transactions
- Unsigned transactions
- Unsigned transactions with signed payload`

>   },

Author: I think we need the bracket, don't we?

>   },

Author: I see. You are suggesting to remove the `Use Cases` paragraph.

  pub mod crypto {
    use super::KEY_TYPE;
    use sp_core::sr25519::Signature as Sr25519Signature;
    use sp_runtime::{
      app_crypto::{app_crypto, sr25519},
      traits::Verify, MultiSignature, MultiSigner
    };
    app_crypto!(sr25519, KEY_TYPE);

    pub struct TestAuthId;

    // implemented for runtime
    impl frame_system::offchain::AppCrypto<MultiSigner, MultiSignature> for TestAuthId {
      type RuntimeAppPublic = Public;
      type GenericSignature = sp_core::sr25519::Signature;
      type GenericPublic = sp_core::sr25519::Public;
    }
  }
  ```

  The [`app_crypto` macro](https://paritytech.github.io/substrate/latest/sp_application_crypto/macro.app_crypto.html) above declares an account with sr25519 signature that is identified by
  `KEY_TYPE`. Note that this doesn't create a new account, but just declaring that a crypto account
  is available for this pallet. We need to initialize this account later. More on this at the
  following.

4. Now, you could send a signed transaction on-chain by first getting a signer.

  ```rust
  fn offchain_worker(block_number: T::BlockNumber) {
    let signer = Signer::<T, T::AuthorityId>::all_accounts();

    // ...
  }
  ```

>   Although it is calling `all_accounts()`, we will only inject one account into this pallet later

Author: I have re-written this sentence.

      //   The returned result means:
      //   - `None`: no account is available for sending transaction
      //   - `Some((account, Ok(())))`: transaction is successfully sent
      //   - `Some((account, Err(())))`: error occured when sending the transaction
      if let Some((_, res)) = signer.send_unsigned_transaction(
        // this line is to prepare and return payload
        |acct| Payload { number, public: acct.public.clone() },
        |payload, signature| Call::some_extrinsics { payload, signature },
      ) {
        match res {
          Ok(()) => log::info!("unsigned tx with signed payload successfully sent.");
          Err(()) => log::error!("sending unsigned tx with signed payload failed.");
        };
      } else {
        // The case of `None`: no account is available for sending
        log::error!("No local account available");
      }
    }
  }
  ```

  In the above code snippet, we fetch the `signer` and then call `send_unsigned_transaction()`.
  It accepts two function closures. The first one returns the payload to be used, and the second
  one returns the on-chain call with payload and signature passed in. This call returns a

>   `Option<(Account<T>, Result<(), ()>)>` result type, which carry the meaning:

Reviewer: ```suggestion
  `Option<(Account<T>, Result<(), ()>)>` result type to allow for the following results:
```

>     * `None`: no account is available for sending transaction

Reviewer: ```suggestion
    * `None` if no account is available for sending the transaction.
```

>     * `Some((account, Ok(())))`: transaction is successfully sent

Reviewer: ```suggestion
    * `Some((account, Ok(())))` if the transaction is successfully sent.
```

>     * `Some((account, Err(())))`: error occured when sending the transaction

Reviewer: ```suggestion
    * `Some((account, Err(())))` if an error occurs when sending the transaction.
```

> 3. Lastly, you may have a more complex `ValidateUnsigned` logic to check the provided signature

Reviewer: ```suggestion
You can implement more complex logic for `ValidateUnsigned` trait, if needed.
```

>   does match with the public key signing the payload. You can do that with the following logics

Reviewer: ```suggestion
  For example, to check whether a provided signature matches the public key used to sign the payload, you can implement the `ValidateUnsigned` trait using code similar to the following:
```

>   in the `ValidateUnsigned` implementation.

Reviewer: ```suggestion
```

  ```rust
  #[pallet::validate_unsigned]
  impl<T: Config> ValidateUnsigned for Pallet<T> {
    type Call = Call<T>;

    fn validate_unsigned(_source: TransactionSource, call: &Self::Call) -> TransactionValidity {
      let valid_tx = |provide| ValidTransaction::with_tag_prefix("ocw-demo")
        .priority(UNSIGNED_TXS_PRIORITY)
        .and_provides([&provide])
        .longevity(3)
        .propagate(true)
        .build();

      match call {
        Call::unsigned_extrinsic_with_signed_payload {
          ref payload,
          ref signature
        } => {
          if !SignedPayload::<T>::verify::<T::AuthorityId>(payload, signature.clone()) {
            return InvalidTransaction::BadProof.into();
          }
          valid_tx(b"unsigned_extrinsic_with_signed_payload".to_vec())
        },
        _ => InvalidTransaction::Call.into(),
      }
    }
  }
  ```

>   We have seen most of the above code before. The new code is that we use

Reviewer: ```suggestion
  This example uses [`SignedPayload`](https://paritytech.github.io/substrate/latest/frame_system/offchain/trait.SignedPayload.html?search=SignedPayload) to verify that the public key in the payload has the same signature as the one provided.
```

  [`SignedPayload`](https://paritytech.github.io/substrate/latest/frame_system/offchain/trait.SignedPayload.html?search=SignedPayload)
  to verify the payload (which contains the public key) has the same signature as the one provided.

> A full example of [the off-chain function call](https://github.com/paritytech/substrate/blob/monthly-2021-12/frame/examples/offchain-worker/src/lib.rs#L505-L532)

Reviewer: ```suggestion
You can see a full example in the [off-chain function call](https://github.com/paritytech/substrate/blob/monthly-2021-12/frame/examples/offchain-worker/src/lib.rs#L505-L532) and [`ValidateUnsigned` implementation](https://github.com/paritytech/substrate/blob/monthly-2021-12/frame/examples/offchain-worker/src/lib.rs#L308-L318).
```

> and [`ValidateUnsigned` implementation](https://github.com/paritytech/substrate/blob/monthly-2021-12/frame/examples/offchain-worker/src/lib.rs#L308-L318)

Reviewer: ```suggestion
```

> can be seen.

Reviewer: ```suggestion
```

---

> Whew! This is a long chapter. But congratulation! By now you should be able to use off-chain workers

Reviewer: ```suggestion
You have now seen how you can use off-chain workers to send data for on-chain storage using:
```

> to send data on-chain using:

Reviewer: ```suggestion
```
